### PR TITLE
Feat/bulk ack

### PR DESCRIPTION
**Pull request by afrancon**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,8 +29,8 @@ In order to add devices to the groups, please follow these steps :
 
 Groups can be added manually or using the group.py script 
 
-2.1 use group.py with export option to get the csv file
-         python3 group.py -export
+2.1 use group.py with export option to get the csv file \
+```python3 group.py -export``` \
 2.2 Add needed groups to the CSV file then use the import option of group.py scrypt
```
Comment (kyechcha) [inline README.md:34]: script (typo)

File: README.md [hunk 2]
```diff
@@ -55,6 +56,26 @@ Groups can be added manually or using the group.py script
 
 ![](CCV-auto-group.png)
 
+6. Acknowledging vulnerabilities in bulk
+
+6.1 --export-affected-devices: use ack_vuln.py with export-affected-devices options to export vulnerable devices into CVS file. This script takes two additional parameters \
```
Comment (kyechcha) [inline README.md:61]: CSV (not CVS)

Also, should we say "This script takes.." or "This operation takes" ? (this comment applies to other operations below as well)

File: README.md [hunk 3]
```diff
@@ -55,6 +56,26 @@ Groups can be added manually or using the group.py script
...
+a. --cve-id [mandatory] \
+b. --export-affected-devices-preset-name  [optional] if this value is not provided, it takes "All data" as default value \
+E.g., ``` python3 ack_vuln.py --export-affected-devices --cve-id=CVE-2014-0317 ```
+
+6.2 --ack-by-preset: use ack_vuln.py with ack-by-preset options to acknowledge vulnerable devices by preset name and cve. This script takes three additional parameters \
+a. --cve-id [mandatory] \
+b. --ack-by-preset-preset-name  [optional] if this value is not provided, it takes "All data" as default value \
+c. --ack-comments [optional] if this value is not provided, it takes default value of "Default acknowledge comment"
```
Comment (kyechcha) [inline README.md:69]: Comment is not optional per last discussion. Please fix.

File: README.md [hunk 4]
```diff
@@ -55,6 +56,26 @@ Groups can be added manually or using the group.py script
...
+E.g., ``` python3 ack_vuln.py --ack-by-preset --cve-id=CVE-2014-0317 --ack-comments="Testing Bulk Ack from Python"```
+
+6.3 --ack-by-devices: use ack_vuln.py with ack-by-devices options to acknowledge vulnerable devices that were exported in first options. This script takes additional parameters \
```
Comment (kyechcha) [inline README.md:72]: exported using the --export-affected-devices option

File: README.md [hunk 5]
```diff
@@ -55,6 +56,26 @@ Groups can be added manually or using the group.py script
...
+a. --cve-id [mandatory] \
+b. --ack-comments [optional] if this value is not provided, it takes default value of "Default acknowledge comment"
```
Comment (kyechcha) [inline README.md:74]: Comment is not optional per last discussion. Please fix.

File: README.md [hunk 6]
```diff
@@ -55,6 +56,26 @@ Groups can be added manually or using the group.py script
...
+ 
+E.g., ``` python3 ack_vuln.py --ack-by-devices --cve-id=CVE-2014-0317 --ack-comments="Testing Bulk Ack from Python"```
+
+Note: Above script does not refresh preset, if you need latest data then kindly refresh preset from GUI and then run above scripts.
```
Comment (kyechcha) [inline README.md:78]: if you need latest data, then, kindly refresh preset from GUI or API and then run the above script.

File: ack_vuln.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+# Version 1.3 - 2023-01-09
+import argparse
+import csv
+from json import dumps
+
+import cvconfig
+import api
+
+
+# TODO: Throw error if cve-id is not provided
```
Comment (kyechcha) [inline ack_vuln.py:13]: Is this TODO still relevant?

File: ack_vuln.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+def main():
+    parser = argparse.ArgumentParser(
+        prog="ack_vuln.py",
+        description="Vulnerabilities Management")
+
+    # Environment Options parsing
+    parser.add_argument("--token", dest="token",
+                        help="Use this token")
+    parser.add_argument("--center-ip", dest="center_ip",
+                        help="Specified the center FQDN or IPv4 address (default:'cybervision')")
```
Comment (kyechcha) [inline ack_vuln.py:23]: Specify

File: ack_vuln.py [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+    parser.add_argument("--center-port", dest="center_port", default=cvconfig.center_port,
+                        help="Specified the center port (default: %d)" % cvconfig.center_port)
```
Comment (kyechcha) [inline ack_vuln.py:25]: Specify

File: ack_vuln.py [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+    command_group = parser.add_mutually_exclusive_group()
+    command_group.add_argument("--export-affected-devices", dest="command_export", default=False,
+                               action="store_true",
+                               help="Export all affected devices of a given preset into a CSV file, If Preset name is not passed then \"All data\" preset is considered as default\n")
+    command_group.add_argument("--ack-by-preset", dest="command_ack_by_preset", default=False,
+                               action="store_true",
+                               help="Update all devices in this preset for given CVE ID, If Preset name is not passed then \"All data\" preset is considered as default\n")
+    command_group.add_argument("--ack-by-devices", dest="command_ack_by_devices", default=False,
+                               action="store_true",
+                               help="Update all devices from a CSV file for given CVE ID\n")
```
Comment (kyechcha) [inline ack_vuln.py:37]: Should this say "Update all devices" (wherein we don't say what we are updating) or "Acknowledge the CVE ID for all devices from a CSV file" ?

File: ack_vuln.py [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+
+    parser.add_argument("--cve-id", dest="cve_id", required=True,
+                        help="Specified the UUID of CVE")
```
Comment (kyechcha) [inline ack_vuln.py:40]: Specify the CVE Identifier (it is not the UUID ?)

File: ack_vuln.py [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+
+    parser.add_argument("--preset-name", dest="preset_name",
+                        help="Preset Name, default is %s" % cvconfig.preset_name)
```
Comment (kyechcha) [inline ack_vuln.py:43]: here we are using cvconfig.preset_name to get the default preset name, but in earlier statements above, we are adding the value All Data directly in the string ?

File: ack_vuln.py [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+
+    parser.add_argument("--ack-comments",  dest="ack_comments",
+                        help="Acknowledge comments")
```
Comment (kyechcha) [inline ack_vuln.py:46]: Acknowledge comment - i.e. singular (as we can give only one comment ?)

File: ack_vuln.py [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+
+    parser.add_argument("--filename", dest="filename", default="vulnerable_devices.csv",
+                        help="Use this filename")
```
Comment (kyechcha) [inline ack_vuln.py:49]: Specify the default value used - (as mentioned during the discussions)

File: ack_vuln.py [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+
+    parser.add_argument("--delimiter", dest="csv_delimiter",
+                        help="CSV file delimiter, default is %s" % cvconfig.csv_delimiter)
+
+    parser.add_argument("--encoding", dest="csv_encoding",
+                        help="CSV file encoding, default is %s" % cvconfig.csv_encoding)
+
+    parser.add_argument("--proxy", dest="proxy", default=cvconfig.proxy,
+                        help="Specified the proxy to use (default: %s)" % cvconfig.proxy)
```
Comment (kyechcha) [inline ack_vuln.py:58]: Specify

File: ack_vuln.py [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+    center_port = set_conf(args.center_port, cvconfig.center_port)
+    csv_encoding = set_conf(args.csv_encoding, cvconfig.csv_encoding)
+    csv_delimiter = set_conf(args.csv_delimiter, cvconfig.csv_delimiter)
+    proxy = set_conf(args.proxy, cvconfig.proxy)
+
+    # Handle default values
+    ack_comments = set_conf(args.ack_comments, cvconfig.ack_comments)
+
+    if not token or not center_ip:
+        print("TOKEN and CENTER_IP are mandatory, check cvconfig.py or us --token/--center-ip")
```
Comment (kyechcha) [inline ack_vuln.py:74]: or use (typo)

File: ack_vuln.py [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+        row['device-riskscore'] = d['riskScore']
+    except KeyError as e:
+        print("KeyError: {} for device {}".format(e, row['device-id']))
+    except Exception as e:
+        print(e)
+
+
+def write_devices(filename, csv_encoding, csv_delimiter, devices):
+    with open(filename, 'w', encoding=csv_encoding) as csvfile:
+        fieldnames = ['device-id', 'device-mac', 'device-ip', 'device-name', 'device-custom-name', 'device-tags', 'device-riskscore']
```
Comment (kyechcha) [inline ack_vuln.py:108]: deivce-tags is present here, but it is not populated in build_device_row ?

File: ack_vuln.py [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+        for d in devices:
+            row = {}
+            build_device_row(row, d)
+            writer.writerow(row)
+        print(f"LOG: Exported {len(devices)} into '{filename}'")
+
+
+def export_vulnerable_devices(center_ip, center_port, token, proxy, filename, csv_delimiter, csv_encoding, cve_id, preset_name):
+    if not preset_name:
+        print(f"WARN: --preset-name not passed, By default \"All data\" preset is considered")
```
Comment (kyechcha) [inline ack_vuln.py:120]: Again, here all data is used directly in the string - should we use the default value from cvconfig.py instead ?

File: ack_vuln.py [hunk 19]
```diff
@@ -0,0 +1,249 @@
...
+        cve_details = api.get_route(session, route)
+        vulnerability_ids = []
+        for cveDetail in cve_details:
+            vulnerability_ids.append(cveDetail['id'])
+
+        if not vulnerability_ids:
+            print(f"ERR: CVE {cve_id} not found")
+            return
+        
+        # hack to get all devices via 'All data' preset, should be removed later
```
Comment (kyechcha) [inline ack_vuln.py:135]: ? removed later

File: ack_vuln.py [hunk 20]
```diff
@@ -0,0 +1,249 @@
...
+            return
+
+        device_details = []
+        for vulnerability_id in vulnerability_ids:
+            route = f"/api/3.0/presets/{preset_id}/visualisations/vulnerability-list/{vulnerability_id}/devices-ids"
+            devices = api.get_route(session, route)
+            if devices:
+                print(f"INFO: device details are being fetched....")
+                for device in devices:
+                    route = f"/api/3.0/devices/"+device
```
Comment (kyechcha) [inline ack_vuln.py:154]: should we use +device this way or :{device} ?

File: ack_vuln.py [hunk 21]
```diff
@@ -0,0 +1,249 @@
...
+                    device_details.append(device_detail)
+
+        # Loop to build devices
+        write_devices(filename, csv_encoding, csv_delimiter, device_details)
+    return
+
+
+def ack_vulnerabilities_by_preset(center_ip, center_port, token, proxy, cve_id, preset_name, ack_comments):
+    if not preset_name:
+        print(f"WARN: --preset-name not passed, By default \"All data\" preset is considered")
```
Comment (kyechcha) [inline ack_vuln.py:165]: Same comment on All data

File: ack_vuln.py [hunk 22]
```diff
@@ -0,0 +1,249 @@
...
+        preset_name = 'All data'
+
+    if not ack_comments:
+        print(f"--ack-comments is mandatory parameter while acknowledging vulnerabilities")
+        print(f"Usage: python3 ack_vuln.py --ack-by-preset --cve-id CVE_ID --preset-name preset_name --ack-comments ack_comments")
```
Comment (kyechcha) [inline ack_vuln.py:170]: ack-comment (as i suppose we can give only one comment)

File: ack_vuln.py [hunk 23]
```diff
@@ -0,0 +1,249 @@
...
+        return
+
+    with api.APISession(center_ip, center_port, token, proxy) as session:
+        presets = api.get_route(session, '/api/3.0/presets')
+        preset_id = 0
+        if not preset_name:
+            preset_name = 'All data'
```
Comment (kyechcha) [inline ack_vuln.py:177]: again, pick from cvyconfig.py

File: ack_vuln.py [hunk 24]
```diff
@@ -0,0 +1,249 @@
...
+            print(f"RESPONSE: Calling [PUT] {route} got error code {ack_vul_response.status_code}")
+            if ack_vul_response.text:
+                print(f"ERROR: Info {ack_vul_response.text}")
+
+    return
+
+
+def ack_vulnerabilities_by_devices(center_ip, center_port, token, proxy, filename, csv_delimiter, cve_id, ack_comments):
+    if not ack_comments:
+        print(f"--ack-comments is mandatory parameter while acknowledging vulnerabilities")
```
Comment (kyechcha) [inline ack_vuln.py:222]: ack comment

File: api.py [hunk 25]
```diff
@@ -118,6 +118,12 @@ def patch_route(session, route, json):
     return session.patch(route,json=json)
 
 
+def put_route(session, route, json):
+    #print(f"DEBUG: [PUT] Calling {route}")
```
Comment (kyechcha) [inline api.py:122]: remove commented lines if we are not using

File: cvconfig.py [hunk 26]
```diff
@@ -18,3 +18,7 @@
 #csv_encoding = 'iso-8859-15'
 csv_encoding = 'utf-8'
 csv_delimiter = ','
+
+# ack_vuln parameters
+preset_name = 'All data'
+ack_comments = ''
```
Comment (kyechcha) [inline cvconfig.py:24]: we don't need ack_comments here - remove.

File: ack_vuln.py (hunk 12 above)
Comment (hsjagadeesh) [reply ack_vuln.py:42]: Updated.

**Pull request merged**